### ISSUE TITLE
`uplink-sys` not building on ubuntu or macos

### ISSUE THREAD

**Issue by eopb** [storj-thirdparty/uplink-rust#65]
Having installed the dependencies listed in [uplink-sys/README.md](https://github.com/storj-thirdparty/uplink-rust/blob/main/uplink-sys/README.md) I am unable to get `uplink-sys` to build on Ubuntu.

```
root@c5b5bf6673cf:/home/uplink-rust# RUST_BACKTRACE=1 cargo c
   Compiling uplink-sys v0.5.2 (/home/uplink-rust/uplink-sys)
error: failed to run custom build command for `uplink-sys v0.5.2 (/home/uplink-rust/uplink-sys)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_DEV_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/home/uplink-rust/target/debug/build/uplink-sys-bea092478836f40f/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-lib=static=uplink
  cargo:rustc-link-search=/home/uplink-rust/target/debug/build/uplink-sys-440350d9c04d666c/out/uplink-c/.build
  cargo:rerun-if-changed=/home/uplink-rust/target/debug/build/uplink-sys-440350d9c04d666c/out/uplink-c/.build/uplink/uplink.h

  --- stderr
  make: *** No rule to make target 'build'.  Stop.
  rm: cannot remove 'uplink-c/.build': No such file or directory
  error: header '/home/uplink-rust/target/debug/build/uplink-sys-440350d9c04d666c/out/uplink-c/.build/uplink/uplink.h' does not exist.
  thread 'main' panicked at 'Error generating bindings.: ()', uplink-sys/build.rs:117:10
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5
     1: core::panicking::panic_fmt
               at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14
     2: core::result::unwrap_failed
               at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/result.rs:1651:5
     3: core::result::Result<T,E>::expect
     4: build_script_build::main
     5: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
root@c5b5bf6673cf:/home/uplink-rust#
```

**Comment by ifraixedes** [comment]
Did you try building it from [repo](https://github.com/storj-thirdparty/uplink-rust/blob/main/uplink-sys/README.md#building-from-repo) or from [crates.io](https://github.com/storj-thirdparty/uplink-rust/blob/main/uplink-sys/README.md#building-from-cratesio)?

**Comment by eopb** [comment]
From the repo. I tried building `main` and https://github.com/storj-thirdparty/uplink-rust/commit/e3802aba90f28c9e514d9f2d929b21ef9913b63b which is the tag for the latest release on crates.io

**Comment by eopb** [comment]
I've just also tried building from crates.io by adding it as a dependency of an empty project with go installed. Using this method it still throws the same error while building `uplink-sys`

**Comment by ifraixedes** [comment]
Thank you for the clarification, I will try to take a look during this week.

**Comment by eopb** [comment]
Here's a `Dockerfile` to help with reproducing the error when you're able to look at it

```dockerfile
from ubuntu:latest

run apt update
run apt install golang-go build-essential libclang-dev rustc cargo git -y

run git clone https://github.com/storj-thirdparty/uplink-rust.git

workdir /uplink-rust/uplink-sys

run cargo check
```

**Comment by ifraixedes** [comment]
I ran a Docker container from the `ubuntu:last` and accessed it through an interactive shell.

I installed the mentioned packages and then I created a new Rust binary project.
I added `uplink-sys` with `cargo add`.
I copied into `main.rs` the content of [`list_buckets` function](https://github.com/storj-thirdparty/uplink-rust.git).
And I ran it successfully.

On the other hand, I cloned the repo and it failed with the error that you reported, but the issue is that the Git submodule to uplink-c was initialized. I initialized (i.e `git submodule update --init --recursive`) and then `cargo build` succeded.

Using the `Dockerfile` works if you add the `--recurse-modules` flag to the `git clone` command.
